#24 Feature: New offline dictionary implementation without need to pad bytes elements (variable length values)

buchireddy opened the pull request:
See #24 for more details.

buchireddy committed:
Adding a variable length bytes value reader for offline dictionary index.

buchireddy committed:
Make variables private.

kishoreg commented on pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/SegmentDictionaryCreator.java:
```diff
@@ -154,24 +155,18 @@ public void build()
         // Backward-compatible: index file is always big-endian
-        try (PinotDataBuffer dataBuffer = PinotDataBuffer
-            .mapFile(_dictionaryFile, false, 0, (long) numValues * _numBytesPerEntry, ByteOrder.BIG_ENDIAN,
+        long size = VarLengthStringsReaderWriter.getRequiredSize(sortedStrings);
```
we need to support both modes.  We need the new implementation for byte and string column and only when the standard deviation is high.

buchireddy replied:
Are you saying we calculate the stdDev and decide to use whether fixed or var length dict based on the std dev on-the-fly?

kishoreg commented on pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java:
```diff
@@ -0,0 +1,141 @@
+import java.io.IOException;
+import org.apache.pinot.core.segment.memory.PinotDataBuffer;
+
+public class VarLengthBytesValueReaderWriter implements Closeable, ValueReader {
```
we just need reader

buchireddy replied:
Well, the `init()` method is actually writing the data into the buffer, hence I've kept the name ReaderWriter.

buchireddy committed:
Cleaned up the changes a bit and removed the hacks to use the
variable length byte dictionary.

In the next iteration, we'll be deciding dynamically whether to use
fixed or variable length impl.

buchireddy committed:
Minor cleanup.

buchireddy committed:
Adding stats about the string values length in the StringDictionaryPerfTest.java

kishoreg commented on pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java:
```diff
@@ -0,0 +1,166 @@
+   * collisions with the regular int/string dictionary values written in fixed size
+   * format.
+   */
+  private static final byte[] MAGIC_HEADER = StringUtil.encodeUtf8("vl1;");
```
What happens when we get to vl10 or come with some other encoding schema say foo1. We would want to keep this backward compatible rt?

buchireddy replied:
I've been thinking about this further too. I'll check how this is being done (if at all) in other parts of Pinot so that I can keep it consistent. 

Here are some rough starter ideas I've:
* Keep the length at 8 bytes leaving the scope for future versions and some other bits as well. That will also allow us to add a couple of other char sequences, making the header (almost) uniquely identifiable.
* Keep the size of the header also variable and add the length first. We'll read the length first and then read those many bytes to get the header. The chances of both the length and header colliding with a fixed size dictionary are very very rare.

Do you have any ideas?

kishoreg commented on pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java:
```diff
@@ -0,0 +1,166 @@
+      length = _dataBuffer.getInt(INDEX_ARRAY_START_OFFSET + Integer.BYTES * (index + 1)) - offset;
+    }
+
+    byte[] b = buffer == null ? new byte[length] :
```
what is the purpose of these two lines? recursive ternary operators make it hard to parse the code.

buchireddy replied:
The callers of this method might pass a null byte buffer too and we create a new buffer here in that case. The caller knows the buffer length in advance for fixed length dictionary hence it can create and pass the buffer but it's not possible for var length dict, hence we create the buffer here.

I'll simplify the code for readability.

buchireddy replied:
This is done.

kishoreg commented on pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/SegmentDictionaryCreator.java:
```diff
@@ -185,22 +177,14 @@ public void build()
         Preconditions.checkState(numValues > 0);
         _bytesValueToIndexMap = new Object2IntOpenHashMap<>(numValues);
 
+        byte[][] sortedByteArrays = new byte[sortedBytes.length][];
```
can we avoid creating this?

buchireddy replied:
We can directly use the `ByteArray[]` but that would duplicate the code in `writeBytesValueDictionary()`. Please note we don't recreate the byte arrays here, just an array of references.

buchireddy replied:
Discussed this with Kishore and explain that we don't recreate the byte arrays, just a new array with references to the same `byte[]` instances.

kishoreg commented on pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/SegmentDictionaryCreator.java:
```diff
@@ -211,6 +195,42 @@ public void build()
+  private void writeBytesValueDictionary(int numValues, byte[][] sortedByteArrays) throws IOException {
+    // TODO: Fix this to be dynamic based on the standard deviation in the lengths
+    // of the values in byte array, max length of array, etc.
+    boolean useFixedSizeDict = false;
```
make this configurable for now and set the default value to true.

buchireddy replied:
Will do.

buchireddy replied:
Done. I've modified the config classes to make this configurable and keep the featured disabled by default.

kishoreg commented on pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java:
```diff
@@ -0,0 +1,166 @@
+  }
+
+  public static boolean isVarLengthBytesDictBuffer(PinotDataBuffer buffer) {
+    byte[] header = new byte[MAGIC_HEADER_LENGTH];
```
we need to make this check stronger. There are a few cases this can fail
- dictionary has only one value and its value length is less than vl1;
- if the dictionary has a value vl1; and it happens to be  the first value after sorting the values

buchireddy replied:
Good points. Will fix the first one and let's discuss more on second point.

buchireddy replied:
Fixed.

buchireddy committed:
Improved the magic header for VariableLength dictionary.

Also fixed other review comments from Kishore.

buchireddy committed:
Making the feature of whether to use variable length dictionary
for a column configurable through TableConfig.

buchireddy committed:
Use the variable length dictionary config in realtime segments path too.

buchireddy committed:
Fixed a bug that was caught by the unit tests.

buchireddy committed:
Adding variable length dictionary config in more places like:
- IndexLoadingConfig
- RealtimeSegmentConfig

buchireddy committed:
Use separate file names for different tests in a single class. This was
causing a race condition and unit test failure.

xiangfu0 committed:
Dont set null value for varLengthDictionaryColumns

buchireddy commented:
FTR, I've done some benchmarking of the new variable length dictionary for strings and compared the numbers with the current FixedLength dictionary and here are the results.

**Time taken to lookup 10M values:**
<img width="882" alt="TimeVsDictionarySizesCharts" src="https://user-images.githubusercontent.com/945283/59487242-edc20280-8e30-11e9-92d7-1891379d4639.png">
**Observation:** As can be clearly seen in the graph, as strings are getting bigger, the variable lengh dictionary is giving much better lookup latencies. When the dictionary size (cardinality) is >1M and the string sizes are small (<100), FixedLength dictionary has better lookup latencies though. 

**Storage requirements of VarLength dictionary:**
Since the variable length dictionary doesn't do any padding, it saves the space for all the cases where the strings in the dictionary aren't of equal length. Hence, this graph plots the % storage savings with VarLength dictionary instead of absolute values.
<img width="612" alt="DictSizeVsStorageSavingsChart" src="https://user-images.githubusercontent.com/945283/59487464-9ff9ca00-8e31-11e9-9e60-13228766662e.png">
**Observation:** If the strings in the dictionary are of different lengths, VarLength dictionary saves 40% space compared to the fixed length dictionary. 

P.S: All raw values from the benchmarking are available at https://docs.google.com/spreadsheets/d/1iOLyhD4AUZw3JsdOkmH6h36KWYalVeUBIVcty6Pnv0E/edit?usp=sharing so feel free to copy/comment on the results.

kishoreg commented on pinot-core/src/main/java/org/apache/pinot/core/indexsegment/generator/SegmentGeneratorConfig.java:
```diff
@@ -275,6 +280,14 @@ public void setColumnSortOrder(List<String> sortOrder) {
+    return _varLengthDictionaryColumns;
+  }
+
+  public void setVarLengthDictionaryColumns(List<String> _varLengthDictionaryColumns) {
```
remove _ for param name

buchireddy replied:
Done.

kishoreg commented on pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java:
```diff
@@ -0,0 +1,196 @@
+    return false;
+  }
+
+  public VarLengthBytesValueReaderWriter(PinotDataBuffer dataBuffer) {
```
move this constructor to the top.

buchireddy replied:
Done.

kishoreg reviewed (approved):
Nicely done!.

buchireddy committed:
Fixed review comments.

buchireddy committed:
Merge branch 'develop' into feature/var_length_bytes

mayankshriv commented on pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java:
```diff
@@ -0,0 +1,196 @@
+  private transient int _numElements;
+
+  public VarLengthBytesValueReaderWriter(PinotDataBuffer dataBuffer) {
+    this._dataBuffer = dataBuffer;
```
With '_', 'this' is redundant.

mayankshriv replied:
I think the convention is to use ReaderWriter for mutable case (realtime indices/dictionaries) and Reader for immutable. Will this be used for mutable case as well?

buchireddy replied:
I agree. Kishore and I have discussed this as well. Since `FixedByteValueReaderWriter` has ReaderWriter in the name, I've named the new class also to keep them consistent. I'll try to rename both in a separate PR.

mayankshriv commented on pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java:
```diff
@@ -0,0 +1,196 @@
+    byte[] b;
+    // If the caller didn't pass a buffer, create one with exact length.
+    if (buffer == null) {
+      b = new byte[length];
```
Would thread-local (reusable) byte array help?

buchireddy replied:
We should be. Let me do that change in another PR.

mayankshriv commented on pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java:
```diff
@@ -0,0 +1,196 @@
+ *
+ * @see FixedByteValueReaderWriter
+ */
+public class VarLengthBytesValueReaderWriter implements Closeable, ValueReader {
```
I am sure there's a reason why you were not able to reuse VarByteChunkSingleValueReader/Writer. Would be good to document that, and may be think if the two can be combined to share some code.

buchireddy replied:
Will add the documentation in the next PR.

mayankshriv commented on pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/ImmutableDictionaryReader.java:
```diff
@@ -244,6 +252,6 @@ protected String getPaddedString(int dictId, byte[] buffer) {
 
   protected byte[] getBuffer() {
-    return new byte[_numBytesPerValue];
+    return (_valueReader instanceof VarLengthBytesValueReaderWriter) ? null : new byte[_numBytesPerValue];
```
Is there a way to avoid 'instanceof'?

buchireddy replied:
If we want to avoid it, it's easily doable. Given that `_valueReader` is a final field, we can have a boolean for this condition and use that here.

mayankshriv commented on pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java:
```diff
@@ -0,0 +1,196 @@
+  }
+
+  @Override
+  public String getUnpaddedString(int index, int numBytesPerValue, byte paddingByte, byte[] buffer) {
```
Both getUnpaddedString() and getPaddedString() have the same code, and seem to return the same value. The callers of these api's might be expected different values.

buchireddy replied:
Correct and IMO, I think the `ValueReader` interface should only have `getString()` rather than padded and unpadded versions but because it has both of those methods, we can't pick/implement only one in VarLengthDictionary. 
Again, let me think more about a better way to clean this up. I wanted to keep the changes to minimal in my very first PR :)

mayankshriv reviewed (commented):
Thanks for doing this, it is a good change. Please consider merging into apache-pinot repo.
